---
documentclass: nojss
author:
  - name: Alex Asher
    affiliation: Texas A\&M University
    address: >
      Department of Statistics
    email: \email{alexasher@stat.tamu.edu}
title:
  formatted: "Semiparametric Analysis of Polygenic Gene-Environment Interactions in Case-Control Studies with \\pkg{caseControlGE}"
  # If you use tex in the formatted title, also supply version without
  plain:     "Semiparametric Analysis of Polygenic Gene-Environment Interactions in Case-Control Studies with caseControlGE"
  # For running headers, if needed
  short:     "\\pkg{caseControlGE}: Gene-Environment Interactions in Case-Control Studies"
abstract: >
  Gene-environment interactions can be efficiently estimated in case-control data by exploiting the assumption of gene-environment independence in the source population, but until recently such techniques required parametric modelling of the genetic variables.  The \pkg{caseControlGE} package implements the methods of \citeauthor{Stalder2017} (2017, \emph{Biometrika}, \textbf{104}, 801-812) and \citeauthor{Wang2018unpublished} (2018, unpublished), which exploit the assumption of gene-environment independence without placing any assumptions on the marginal distributions of the genetic and environmental variables.  These methods are ideally suited for analysis of complex polygenic data for which parametric distributional models are not feasible.  In addition to the two estimators, the package also supplies a function to simulate case-control data and several helper functions for use on model objects.  Use of this package is illustrated using simulated data from a case-control study of breast cancer.
keywords:
  # at least one keyword must be supplied
  formatted: [case-control study; gene-environment interaction; genetic epidemiology; retrospective method; semiparametric analysis; pseudolikelihood; polygenic analysis]
  plain: [case-control study; gene-environment interaction; genetic epidemiology; retrospective method; semiparametric analysis; pseudolikelihood; polygenic analysis]
preamble: >
  \usepackage{amsmath}
  <!-- \usepackage{pdfpages} -->
  <!-- \usepackage[letterpaper]{geometry} -->
output: rticles::jss_article
vignette: >
  %\VignetteIndexEntry{Semiparametric Analysis of Polygenic Gene-Environment Interactions in Case-Control Studies with caseControlGE}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
\begin{titlepage}
\end{titlepage}

\def\bbeta{\mbox{\boldmath $\beta$}}
\def\pr{\hbox{pr}}

```{r setup, cache=FALSE, include=FALSE}
knitr::opts_chunk$set(
  collapse=TRUE, cache=TRUE, autodep=TRUE
  # , echo=TRUE
  # , comment = "#>"
)
```

\section{Introduction}
The \pkg{caseControlGE} package \citep{Asher2018R} contains tools for the analysis of case-control data using \proglang{R} \citep{R2018}.  It implements the methods of \citet{Stalder2017} and \citet{Wang2018unpublished}, both of which fall under the class of semiparametric retrospective profile likelihood estimators.  These methods are the first available to exploit the assumption of gene-environment independence while treating the genetic component nonparametrically.  As such, they are well suited to replace logistic regression as the preferred method in situations where parametric distributional models are not feasible, such as in the analysis of complex polygenic data.

An important aspect of case-control studies is that the covariates are sampled conditional on the response, disease status.  Given the genetic and environmental covariates $G$ and $E$, we assume the risk of disease $D$ in the underlying population follows the model \[ \pr(D=1 \mid G,X) = H\{\alpha_0 + m(G,X,\bbeta)\}, \] where $H(x)=\{ 1 + \exp(-x)\}^{-1}$ is the logistic distribution function and $m(G,X,\bbeta)$ is a function that describes the joint effect of $G$ and $X$ and is known up to the unspecified parameters of interest $\bbeta$.  

Given the retrospective nature of case-control sampling, it is surprising that standard prospective logistic regression can be used to obtain unbiased estimates of $\bbeta$ \citep{PrenticePyke1979}.  Logistic regression requires no assumptions about the joint distribution of $G$ and $E$, but it suffers from low power when estimating $G * E$ interaction effects.  To gain efficiency, \citet{ChatterjeeCarroll2005} exploited the assumption of gene-environment independence in the source population to maximize the retrospective likelihood while profiling out the distribution of $E$.  Their method is available as the function \code{snp.logistic} in the \emph{Bioconductor} package \pkg{CGEN} \citep{CGEN2012}.

The method of \citeauthor{ChatterjeeCarroll2005}, and subsequent methods utilizing the same retrospective profile likelihood framework, require a parametric model for the distribution of $G$ given $E$.  This becomes difficult as the number and complexity of genetic variables in the model grows.  Capitalizing on advances in high-throughput genomics, genome-wide association studies have identified scores of SNPs associated with complex diseases such as cancers and diabetes.  Modern case-control studies of gene-environment interactions need efficient methodology that allows for a flexible and arbitrarily complex genetic component, such as multiple correlated SNPs and/or continuous polygenic risk scores (PRSs).

The method of \citet{Stalder2017} extends the retrospective profile likelihood framework of \citeauthor{ChatterjeeCarroll2005}, dispensing with the need to model $G$ parametrically.  When the population disease rate $\pi_1$ is known, the retrospective profile loglikelihood can be estimated (up to an additive constant) using just the case-control sample and without modeling the distribution of $G$.  When $\pi_1$ is unknown but the disease is rare, estimates can be obtained using the \emph{rare disease approximation} that $\pi_1 \approx 0$, which typically introduces negligable bias \citep{Stalder2017}.

The semiparametric method of \citet{Stalder2017} is implemented 


-

-



These methods extend the retrospective profile likelihood framework developed by \citet{ChatterjeeCarroll2005}, which addresses the retrospective sampling scheme of case-control studies by 
which gains efficiency over logistic regression by exploiting the assumption of gene-environment independence in the population.  The method of \citeauthor{ChatterjeeCarroll2005} 

\pkg{caseControlGE} contains three main functions: \code{simulateCC}, \code{spmle}, and \code{spmleCombo}, as well as several helper functions.  Section \ref{sec:simulateCC} of this paper introduces \code{simulateCC} in the context of simulating case-control data analogous to the data analyzed in \cite{Wang2018unpublished}.  Section \ref{sec:spmle} introduces \code{spmle} as a tool to analyze the simulated data, and section \ref{sec:spmleCombo} introduces \code{spmleCombo} to conduct a more efficient analysis of the simulated data.

\section{Simulating case-control data with simulateCC} \label{sec:simulateCC}
\cite{Wang2018unpublished} demonstrate the utility of their method 

 

```{r}
1
```

\section{Analyzing case-control data with spmle} \label{sec:spmle}


```{r}
1
```

\section{Analyzing case-control data with spmleCombo} \label{sec:spmleCombo}


\bibliography{RJCrefs_05-30-2018}


<!-- \includepdf[pages=-, fitpaper=true]{/home/alex/Research/Tianying/caseControlGE/caseControlGE} -->
